### ISSUE TITLE
Upgrade System Fix

### ISSUE THREAD

**Issue by zalberico** [urbit/urbit#5897]
The last kelvin change OTA led to multiple issues: either apps would fail to revive after upgrade or fail to suspend/upgrade.

This led to issues where users would get stuck on an old version with confusing notifications about suspended apps or apps blocking upgrades.

The fix required manually suspending apps, installing the OTA, and then reviving apps via the dojo - obviously not good for zero click maintenance.

The locus of improvements required to fix this exist in clay, gall, and kiln. Ted and Phililp are writing the fix for this on philip/agent-clay. We need this fix in prior to the next kelvin change release (remote scry OTA).

https://gist.github.com/belisarius222/2ae74bfb5a40860b59d28970d29b3329

**Comment by zalberico** [comment]
We've hit this now (hood failing to start) while running an upgrade for the gall request queue fix which is not a kelvin version change.

This means we're blocked on the upgrade fix for us to be able to deploy the gall request queue change work safely. This is the most important issue/bottleneck blocking releases.

**Comment by zalberico** [comment]
Ran into unexpected issue last week that had to do with move ordering.
 - Philip worked it through with Ted and have an implemented solution that they think solves this.
 - Back to testing the change (prerequisite for making sure things are good to merge).
 - Philip came to the conclusion we should be using breadth first move ordering everywhere (this can also go into this release unless we hit more concurrency problems). There's a separate PR for this #6041.

**Comment by zalberico** [comment]
@philipcmonk just following up on this one, let me know when you think it'll be good to get on the pre-release moon.

@arthyn , @jamesacklin: @lukestiles mentioned he'd talk to you about getting a front-end person to assist the bits that @philipcmonk needs for this. This fix is blocking us shipping kelvin version upgrades so it's the most important current thing we need to ship kernel side.

**Comment by tomholford** [comment]
Heard from @lukestiles and @zalberico that @philipcmonk could use a hand with the corresponding frontend changes. Will follow up with him on the network

**Comment by philipcmonk** [comment]
Basically, the old "arak" type has been replaced, and we need to replace its use in the frontend.  @liam-fitzgerald added a "pikes" scry to kiln to replace the information we were getting as araks: 5eb499186d03d34d3cb5e93d0fb3ea84e3abf61e

There aren't too many places we use them, here are the ones I've noticed:

- A few uses in this file, surrounding the kelvin upgrade notification. https://github.com/urbit/urbit/blob/301158c40a2c0639f38e0aaac14eaf48bf0f3568/pkg/grid/src/nav/notifications/SystemNotification.tsx#L42
- Here as well https://github.com/urbit/urbit/blob/301158c40a2c0639f38e0aaac14eaf48bf0f3568/pkg/grid/src/nav/preferences/AppPrefs.tsx#L13

Additionally, kiln-bump (the action that happens when you click the button to force a kelvin upgrade) now takes no arguments (ie only the argument `~`).

Work is on the philip/agent-clay branch.  I recommend starting a ship with the multi-brass pill (`urbit -F zod -B ~/urbit/bin/multi-brass.pill`) to get something with other desks (only garden and landscape right now).

**Comment by zalberico** [comment]
@tomholford and @philipcmonk - how are things looking on this one? Are we on track for getting this on an internal release moon by EOW? Blocked on anything?

**Comment by tomholford** [comment]
> @tomholford and @philipcmonk - how are things looking on this one? Are we on track for getting this on an internal release moon by EOW? Blocked on anything?

Thanks for checking in. Paired with @liam-fitzgerald to go over the existing changes and the scope of remaining work; and later paired with @ryjm to resolve an issue with the `%pikes` scry (in-flight branch is [here](https://github.com/urbit/urbit/tree/t/5897-arak-to-pike)). Now focused on the corresponding frontend changes. EOW is a stretch but could be possible. Will post a status update after another dev cycle.

**Comment by tomholford** [comment]
@philipcmonk Is `zest.pike` in the `%held` state logically equivalent to `paused.rail.arak`?

**Comment by tomholford** [comment]
@philipcmonk Have updated the `SystemNotifications` logic too. What is the best way to test the updated behavior? In other words, is there a way to get a fake ship into a "base blocked" state?

**Comment by tomholford** [comment]
@philipcmonk looking at `pkg/arvo/lib/hood/kiln.hoon`, not seeing an `on-watch` arm, but I do see an arm called `peer` with the following:

```hoon
++  peer
  |=  =path
  ?>  (team:title our src)
  ?:  =(0 1)  abet  ::  avoid mint-vain
  ?+    path  ~|(kiln-path/path !!)
      [%vats ~]
    (mean leaf+"kiln: old subscription to /kiln/vats failed" ~)
  ==
```

Does this mean that there are no longer any subscriptions necessary for kiln?

**Comment by tomholford** [comment]
@zalberico FYI opened a PR here: https://github.com/urbit/urbit/pull/6049

**Comment by philipcmonk** [comment]
`zest` of `%live` means the app is running, it does not say whether it's receiving OTAs from anywhere.

`%held` means the app is not running but is trying to run.  You get into this state in two main ways: when you're installing an app but it hasn't finished downloading (or it did but failed to install for some reason), or when you forced a kelvin upgrade by suspending desks.

If you change the sys.kelvin of %base to 417 and the `++  zuse  %418` line in sys/zuse.hoon while having other desks that have not yet updated their sys.kelvin, you should see the `SystemNotification` flow.  To test the full flow, you'll want to update the sys.kelvin for %garden as well but not %landscape, so that you can verify the forcing button works.

Yes, there is no more %kiln subscription for vats.  Do you think we need one for pike?

**Comment by tomholford** [comment]
@philipcmonk:
> `zest` of `%live` means the app is running, it does not say whether it's receiving OTAs from anywhere.
> 
> `%held` means the app is not running but is trying to run. You get into this state in two main ways: when you're installing an app but it hasn't finished downloading (or it did but failed to install for some reason), or when you forced a kelvin upgrade by suspending desks.

Great, thanks for the explanation; I have updated the [`zest.pike` comment](https://github.com/urbit/urbit/pull/6049/commits/d99dc3e3abc0a808d051e901521e318c04cf023a#diff-4cb6233435ca782b8814f1aaa93ace544515697ea0a6f433a8ef918f9795330aR193) accordingly.

If I understand correctly, this means that there is no analog to the `paused` state previously exposed by `rail.arak`? If so, is there an alternative way to discern whether or not OTAs are enabled for a given app? The goal is replace this snippet from the previous implementation:

```ts
const otasEnabled = !vat?.arak.rail?.paused;
```

> If you change the sys.kelvin of %base to 417 and the ++ zuse %418 line in sys/zuse.hoon while having other desks that have not yet updated their sys.kelvin, you should see the SystemNotification flow. To test the full flow, you'll want to update the sys.kelvin for %garden as well but not %landscape, so that you can verify the forcing button works.

Makes sense, thank you.

>Yes, there is no more %kiln subscription for vats. Do you think we need one for pike?

Still familiarizing myself with the specifics (will have a better grasp after testing), but I think It potentially could be useful in the upgrade flow. If the user forces an update, it would be helpful to provide UI feedback on the transitional states of the process. e.g., click the button to force, show a spinner, then when an update event is published to the client we can render an "upgrade succeeded" message.

Otherwise I can't think of a good reason besides being consistent with other full stack apps like Groups 2. In which case this pattern might be overkill for now, and we can instead optimistically mutate client-side state instead of waiting for an update event.

**Comment by tomholford** [comment]
@philipcmonk:

> If you change the sys.kelvin of %base to 417 and the `++ zuse %418` line in sys/zuse.hoon while having other desks that have not yet updated their sys.kelvin, you should see the `SystemNotification` flow. To test the full flow, you'll want to update the sys.kelvin for %garden as well but not %landscape, so that you can verify the forcing button works.

Tried this out; turns out that in the `philip/agent-clay` branch (which the frontend branch is based on) already had kelvins bumped to 417 in sys.kelvin and sys/zuse.hoon. So, I tried bumping them to 416 and `|commit`ting %base.  Now seeing this:

```
clay: wait-for-kelvin, [need=[%zuse %417] have={[lal=%zuse num=416]}]
```

Is there anything else I need to change or commit before triggering this flow?

**Comment by philipcmonk** [comment]
> If I understand correctly, this means that there is no analog to the paused state previously exposed by rail.arak?

+pike has a `sync` item.  If that's null, it is not receiving OTAs from anywhere, and if it's non-null, then it is.  There is no longer a way to say "I'm configured to receive OTAs from this ship/desk but those are paused".  A desk can be running/suspending (zest) and receiving/not receiving updates (sync), and those states are now completely independent of each other.

https://github.com/urbit/urbit/blob/d99dc3e3abc0a808d051e901521e318c04cf023a/pkg/base-dev/sur/hood.hoon#L5

> I tried bumping them to 416 and |committing %base. Now seeing this:

If you're seeing that, then in grid you should see a notification saying that the upgrade is blocked.  IIRC you might have to click the notification button though, it might not add a number.

**Comment by philipcmonk** [comment]
The lack of notification is also a regression from an earlier version of this branch.  I've identified the issue and will fix it, but in the meantime if you run `|pass [%c %stir %goad ~]` after seeing the `wait-for-kelvin`, the notification should arrive.

**Comment by tomholford** [comment]
> The lack of notification is also a regression from an earlier version of this branch. I've identified the issue and will fix it, but in the meantime if you run `|pass [%c %stir %goad ~]` after seeing the `wait-for-kelvin`, the notification should arrive.

This did the trick!

![image](https://user-images.githubusercontent.com/16504501/200471278-369c2064-f269-4c4c-a085-a83e427101c6.png)

**Comment by tomholford** [comment]
Updated the flows for syncing / unsyncing both app-level and system-level updates. Now instead of a pausable toggle, there's either an 'Unsync' button, or an input to enter a source ship with which to sync. Here's an example:

https://user-images.githubusercontent.com/16504501/200498238-aea08f8f-bb71-4f10-b9fe-e7270bedfe04.mp4

Now going to continue testing the base blocked flow.

**Comment by tomholford** [comment]
@zalberico Status update: PR is ready for review

**Comment by zalberico** [comment]
Thanks @tomholford! @philipcmonk it'd be good to get your branch up into a PR too (even if just WIP)

**Comment by zalberico** [comment]
Pairing with @Fang- today to get this on ready and on the canary. Once we have that we need to communicate out with developers to make the (small) necessary changes on their app.

**Comment by zalberico** [comment]
@Fang- @philipcmonk - were you guys able to pair on this? How are things looking? If you're able to get a PR up I can link it here.

**Comment by Fang-** [comment]
We got it to the point where @philipcmonk can just start running through his tests. Feels really quite close.

**Comment by zalberico** [comment]
Things are looking good here - we want to have this on the canary today.

Steps:
 - Create PR for this
 - Get +2 on PR
 - Merge into next/arvo
 - Deploy tip of next/arvo to canary moon
 - Test on canary moon
 - If looks good tag release candidate
 - Zach to send email internally
 - Deploy to internal moon

**Comment by zalberico** [comment]
The PR was branched for release and deployed on the canary and then on philip's ship. He's currently debugging a road related memory issue. When things look good there, we're going to email the company the steps they need to do after their ship picks up the OTA, then deploy the release candidate to the tlon internal moon. 

If everyone is running successfully for the rest of this week, we'll prepare for general release next Tuesday. App developers will need to be notified to update their apps for the new kelvin version.

**Comment by zalberico** [comment]
@philipcmonk we may want to merge #6109 in for the release too.

**Comment by tomholford** [comment]
> * The App Info in landscape no longer shows relevant info (@tomholford do you have an idea what might cause this?):

Have an idea on what might be the issue here, will investigate and open a PR once I get my ship back up

**Comment by zalberico** [comment]
More seriously, groups 2 and talk were updated to the new kelvin version, but I didn't get a notification for it and they did not automatically revive themselves. @arthyn thinks this is related to %notify.

**Comment by tomholford** [comment]
> * The App Info in landscape no longer shows relevant info (@tomholford do you have an idea what might cause this?):
> * <img alt="Screenshot 2022-11-30 at 8 26 09 AM" width="613" src="https://user-images.githubusercontent.com/441466/204853299-7e38ecdc-3364-4a56-a795-548a8b4e3747.png">

@zalberico: Following up on the missing AppInfo. The data that's missing is from `treaties`. I am reliably able to reproduce that this behavior is working as expected on pre-release ships (treaties are loading, app info is rendering), and broken on post-release (for some reason, the treaty data is not loading). 

Reviewed [the diff](https://github.com/urbit/urbit/compare/master...next/417) between `next/417` and `master`, at this point nothing jumps out at me that would be causing this  observed behavior. It appears that none of the backend or frontend logic was changed for this flow. 

Discussed briefly with @philipcmonk, one thought he had that could be a thread to start pulling on:
>maybe having %garden suspended for a little bit causes problems?

Continuing to investigate.

**Comment by tomholford** [comment]
@zalberico One difference from the screenshot you shared, is that on my test moon the `pike` data (source of  `hash` and `installed into`) is loading as expected:

![image](https://user-images.githubusercontent.com/16504501/204967229-1b3decb4-4895-4396-983b-84ce683f09db.png)

Can you confirm whether or not this is still not loading for you?

**Comment by zalberico** [comment]
They are loading now - this may have been related to me having the wrong %garden version. Thankfully may not be a bug.

**Comment by tomholford** [comment]
> They are loading now - this may have been related to me having the wrong %garden version. Thankfully may not be a bug.

Great! Are all expected properties loading now, or just the "desk hash" and "installed into" fields?

**Comment by zalberico** [comment]
@tomholford all fields I think look good now.

@philipcmonk there's something weird here I think we need to understand the cause of: #6118

**Comment by zalberico** [comment]
417 was shipped to the live network
